Src/Base/AMReX_VisMF.cpp

```diff
@@ -2471,6 +2518,828 @@ VisMF::WriteAsync (const FabArray<FArrayBox>& mf, const std::string& mf_name)
+                            k = gidx[rank][lidx];
+                            ++lidx;
+                        } else {
+                            ++rank;
+                            lidx = 0;
+                        }
+                    } while (k < 0);
+
+                    h.m_min[k].resize(ncomp);
+                    h.m_max[k].resize(ncomp);
+
+                    if (nbytes_on_rank[rank] < 0) { // First time for this rank
+                        std::memcpy(&(nbytes_on_rank[rank]), pgd, sizeof(int64_t));
+                        pgd += sizeof(int64_t);
+                    }
+
+                    int64_t nbytes;
+                    std::memcpy(&nbytes, pgd, sizeof(int64_t));
+                    pgd += sizeof(int64_t);
+
+                    for (int icomp = 0; icomp < ncomp; ++icomp) {
+                        Real cmin, cmax;
+                        std::memcpy(&cmin, pgd             , sizeof(Real));
+                        std::memcpy(&cmax, pgd+sizeof(Real), sizeof(Real));
+                        pgd += sizeof(Real)*2;
+                        h.m_min[k][icomp] = cmin;
+                        h.m_max[k][icomp] = cmax;
+                        h.m_famin[icomp] = std::min(h.m_famin[icomp],cmin);
+                        h.m_famax[icomp] = std::max(h.m_famax[icomp],cmax);
+                    }
+
+                    auto info = StaticWriteInfo(rank);
+                    int fno = std::get<0>(info);   // file #
+                    h.m_fod[k].m_name = amrex::Concatenate(VisMF::BaseName(mf_name)+FabFileSuffix, fno, 5);
+                    h.m_fod[k].m_head = nbytes;
+                }
+            }
+
+            Vector<int64_t> offset(nprocs);
+            for (int ip = 0; ip < nprocs; ++ip) {
+                auto info = StaticWriteInfo(ip);
+                int sno = std::get<1>(info);
+                if (sno == 0) {
+                    offset[ip] = 0;
+                } else {
+                    offset[ip] = offset[ip-1] + nbytes_on_rank[ip-1];
+                }
+            }
+
+            for (int k = 0; k < n_global_fabs; ++k) {
+                h.m_fod[k].m_head += offset[dm[k]];
+            }
+
+            VisMF::WriteHeaderDoit(mf_name, h);
+        }
+
+        int barrier_count = 0;
+        int ranks_in_file = ParallelDescriptor::NProcs(io_comm);
+
+        Vector<MPI_Request> waiting_reqs(ispot);
+        Vector<MPI_Status> waiting_status(ispot);
+        Vector<MPI_Request> syncing_reqs(ranks_in_file - ispot);
+        Vector<MPI_Status> syncing_status(ranks_in_file - ispot);
+
+        Real t0 = amrex::second();
+
+        // If not the first MPI writing on this rank,
+        // block until it is your turn.
+        while (barrier_count < ispot)
+        {
+            waiting_reqs[barrier_count++] =
+                ParallelDescriptor::Abarrier(io_comm).req();
+        }
+        barrier_count = 0; // reset while waiting.
+        ParallelDescriptor::Waitall(waiting_reqs, waiting_status);
```

Comment:
Maybe we should only call this if `ispot > 0`.  It seems that MPI standard does not say what to do if the size of waiting_reqs is zero.